src/compiler/checker.ts

```diff
@@ -13508,7 +13508,12 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             // we substitute an instantiation of E where P is replaced with X.
             return substituteIndexedMappedType(type.objectType as MappedType, type.indexType);
         }
-        const indexConstraint = getSimplifiedTypeOrConstraint(type.indexType);
+        const indexType = type.indexType;
+        let indexConstraint = getSimplifiedTypeOrConstraint(indexType);
+        if (indexConstraint && indexConstraint.flags & TypeFlags.Index) {
+            const constraint = getBaseConstraintOfType((indexConstraint as IndexType).type);

```

Comment:
Hm, by using `getBaseConstraintOfType` here like this, you're shortcutting a lot of steps in `structuredTypeRelatedToWorker`. That's what the baseline changes here show. Specifically, presume you have a `T[keyof U]` being related to a `A` where `A extends T[keyof T]` and `U extends T` - this will cause us to skip past the `T[keyof T]` constraint for `T[keyof U]` straight to `T[string | number | symbol]` (assuming `T` is unconstrained), thereby missing that `T[keyof U]` is an allowed subset of `A` because `T[keyof T]` is `A`'s constraint (under our current rules - read how we have lax indexed access constraint relationships elsewhere). This is, unfortunately, almost definitely a visible breaking change as-is. Rather than skipping so many steps here, explicitly adjusting `structuredTypeRelatedToWorker` to do some _extra_ constraint exploration for `IndexedAccess` types within the `sourceFlags & TypeFlags.TypeVariable` section would be preferable and *probably* less likely to be breaky if it's just an alternate non-error-reporting success path.

Reply:
Process in `sourceFlags & TypeFlags.TypeVariable` section, so that existing error.txt is not affected !